src/menu/index.tsx

```diff
@@ -0,0 +1,276 @@
+	onFocus?(): void;
+	/** Callback called when menu root is blurred */
+	onBlur?(): void;
+	/** Property to determine how many items to render. Not passing a number will render all results */
+	itemsInView?: number;
+	/** Custom renderer for item contents */
+	itemRenderer?(properties: ItemRendererProperties): RenderResult;
+}
+
+export interface ItemRendererProperties {
+	value: string;
+	label?: string;
+	disabled: boolean;
+	active: boolean;
+	selected: boolean;
+}
+
+interface MenuICache {
+	value: string;
+	initial: string;
+	activeIndex: number;
+	numberInView: number;
+	menuHeight: number;
+	itemHeight: number;
+	itemToScroll: number;
+}
+
+const offscreenHeight = (dnode: RenderResult) => {
+	const r = renderer(() => dnode);
+	const div = global.document.createElement('div');
+	div.style.position = 'absolute';
+	global.document.body.appendChild(div);
+	r.mount({ domNode: div, sync: true });
+	const dimensions = div.getBoundingClientRect();
+	global.document.body.removeChild(div);
+	return dimensions.height;
+};
+
+const factory = create({
+	icache: createICacheMiddleware<MenuICache>(),
+	focus,
+	dimensions,
+	cache,
+	theme
+}).properties<MenuProperties>();
+
+export const Menu = factory(function({
+	properties,
+	middleware: { icache, cache, focus, dimensions, theme }
+}) {
+	const {
+		options,
+		initialValue,
+		onValue,
+		onRequestClose,
+		onActiveIndexChange,
+		activeIndex,
+		focusable = true,
+		onBlur,
+		onFocus,
+		itemsInView: numberInView,
+		itemRenderer,
+		theme: themeProp
+	} = properties();
+
+	if (initialValue !== undefined && initialValue !== icache.get('initial')) {
+		icache.set('initial', initialValue);
+		icache.set('value', initialValue);
+		icache.set('activeIndex', findIndex(options, (option) => option.value === initialValue));
+	}
+
+	if (numberInView && numberInView !== icache.get('numberInView')) {
+		icache.set('numberInView', numberInView);
+
+		const offscreenItemProps = {
+			selected: false,
+			onSelect: () => {},
+			active: false,
+			onRequestActive: () => {},
+			onActive: () => {},
+			scrollIntoView: false
+		};
+
+		const itemHeight = icache.getOrSet(
+			'itemHeight',
+			offscreenHeight(
+				<MenuItem {...offscreenItemProps}>
+					{itemRenderer
+						? itemRenderer({
+								selected: false,
+								active: false,
+								value: 'offscreen',
+								disabled: false
+						  })
+						: 'offscreen'}
+				</MenuItem>
+			)
+		);
+		itemHeight && icache.set('menuHeight', numberInView * itemHeight);
+	}
+
+	const selectedValue = icache.get('value');
+	const computedActiveIndex =
+		activeIndex === undefined ? icache.getOrSet('activeIndex', 0) : activeIndex;
+
+	function _setActiveIndex(index: number) {
+		if (onActiveIndexChange) {
+			onActiveIndexChange(index);
+		} else {
+			icache.set('activeIndex', index);
+		}
+	}
+
+	function _setValue(value: string) {
+		icache.set('value', value);
+		onValue(value);
+	}
+
+	function _onKeyDown(event: KeyboardEvent) {
```

Comment:
Do we need to have an underscore prefix? It was originally to indicate private functions on a class, but this is not a class widget.